hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/BackpressureRegulator.java

```diff
@@ -0,0 +1,214 @@
+            throw new IllegalArgumentException("Can't have '"
+                    + props.BACKPRESSURE_SYNCWINDOW.getName() + "' with a value smaller than 0");
+        }
+        return syncWindow;
+    }
+
+    private int getBackoffTimeoutMs(GroupProperties props) {
+        int backoffTimeoutMs = props.BACKPRESSURE_BACKOFF_TIMEOUT_MILLIS.getInteger();
+        if (enabled && backoffTimeoutMs < 0) {
+            throw new IllegalArgumentException("Can't have '"
+                    + props.BACKPRESSURE_BACKOFF_TIMEOUT_MILLIS.getName() + "' with a value smaller than 0");
+        }
+        return backoffTimeoutMs;
+    }
+
+    private int getMaxConcurrentInvocations(GroupProperties props) {
+        GroupProperties.GroupProperty invocationsPerPartitionProp = props.BACKPRESSURE_MAX_CONCURRENT_INVOCATIONS_PER_PARTITION;
+
+        int invocationsPerPartition = invocationsPerPartitionProp.getInteger();
+        if (invocationsPerPartition < 1) {
+            throw new IllegalArgumentException("Can't have '"
+                    + invocationsPerPartitionProp.getName() + "' with a value smaller than 1");
+        }
+
+        return (props.PARTITION_COUNT.getInteger() + 1) * invocationsPerPartition;
+    }
+
+    /**
+     * Checks if back-pressure is enabled.
+     * <p/>
+     * This method is only used for testing.
+     */
+    boolean isEnabled() {
+        return enabled;
+    }
+
+    int getMaxConcurrentInvocations() {
+        if (enabled) {
+            return maxConcurrentInvocations;
+        } else {
+            return Integer.MAX_VALUE;
+        }
+    }
+
+    public CallIdSequence newCallIdSequence() {
+        if (enabled) {
+            return new CallIdSequence.CallIdSequenceWithBackpressure(maxConcurrentInvocations, backoffTimeoutMs);
+        } else {
+            return new CallIdSequence.CallIdSequenceWithoutBackpressure();
+        }
+    }
+
+    /**
+     * Checks if a sync is forced for the given BackupAwareOperation.
+     * <p/>
+     * Once and a while for every BackupAwareOperation with one or more async backups, these async backups are transformed
+     * into a sync backup.
+     *
+     * For {@link com.hazelcast.spi.UrgentSystemOperation} no sync will be forced.
+     *
+     * @param backupAwareOp the BackupAwareOperation to check
+     * @return true if a sync needs to be forced, false otherwise.
+     */
+    public boolean isSyncForced(BackupAwareOperation backupAwareOp) {
+        if (disabled) {
+            return false;
+        }
+
+        // if there are no asynchronous backups, there is nothing to regulate.
+        if (backupAwareOp.getAsyncBackupCount() == 0) {
+            return false;
+        }
+
+        Operation op = (Operation) backupAwareOp;
+
+        // we never apply back-pressure on urgent operations.
+        if (op.isUrgent()) {
+            return false;
+        }
+
+        AtomicInteger syncDelayCounter = syncDelayCounter(op);
+
+        for (; ; ) {
```

Comment:
I'm not worried about this not being wait free because:
- a thread needs to do so much other stuff to complete a backup, that there is a very small chance that 2 threads at the same moment are going to content with each other
- there are 271 partitions by default and each partition will get its own sync delay counter. This means that there is a very small chance that 2 threads are going to content for the same counter.